### PR TITLE
hydra2: Base implementation of timeout functionality

### PR DESCRIPTION
**Pull request by RozanovAnatoliy**
Calling alarm function to get SIGALRM signal if timed out and sending signal to proxies to kill application's processes. Also time left is passed to HYD_bstrap_setup() to check timed out in startup.

### CODE REVIEW

File: src/pm/hydra2/mpiexec/mpiexec.c [hunk 1]
```diff
@@ -835,14 +847,23 @@ int main(int argc, char **argv)
...
         HYD_bstrap_setup(mpiexec_params.base_path, mpiexec_params.launcher,
                          mpiexec_params.launcher_exec, pg->node_count, pg->node_list, -1,
                          mpiexec_params.port_range, args, 0, &pg->num_downstream,
                          &pg->downstream.fd_stdin, &pg->downstream.fd_stdout_hash,
                          &pg->downstream.fd_stderr_hash, &pg->downstream.fd_control_hash,
                          &pg->downstream.proxy_id, &pg->downstream.pid, mpiexec_params.debug,
-                         mpiexec_params.tree_width);
+                         mpiexec_params.tree_width, time_left);
+
+    if (status == HYD_ERR_TIMED_OUT) {
```
Comment (raffenet) [inline src/pm/hydra2/mpiexec/mpiexec.c]: This block reads odd to me. We check for timeout, and then call `HYD_check_timed_out` presumably to print the timeout message?

File: src/pm/hydra2/mpiexec/mpiexec.c [hunk 2]
```diff
@@ -928,14 +949,20 @@ int main(int argc, char **argv)
...
     /* Print exitcodes if necessary */
     if (mpiexec_params.print_all_exitcodes) {
         HYD_PRINT(stdout, "Exit codes: ");
     }
-    exit_status = 0;
+
     for (i = 0; i < pg->num_downstream; i++) {
         char *curr_nodename = NULL;
 
+        if (mpiexec_params.timeout.timed_out && !exitcodes) {
```
Comment (raffenet) [inline src/pm/hydra2/mpiexec/mpiexec.c]: A comment here would be good to indicate we do not need to collect all the exitcodes in a timeout scenario.

File: src/pm/hydra2/mpiexec/mpiexec.c [hunk 3]
```diff
@@ -87,19 +93,23 @@ static void signal_cb(int signum)
...
      * it as a SIGKILL (user convenience to force kill processes). */
-    if (signum == SIGINT && ++sigint_count > 1)
+    if (signum == SIGINT && ++sigint_count > 1) {
         exit(1);
-    else if (signum == SIGINT) {
+    } else if (signum == SIGINT) {
         /* First Ctrl-C */
         HYD_PRINT(stdout, "Sending Ctrl-C to processes as requested\n");
         HYD_PRINT(stdout, "Press Ctrl-C again to force abort\n");
+    } else if (signum == SIGALRM) {
```
[the remark this answers is not in the document]
Comment (RozanovAnatoliy) [reply src/pm/hydra2/mpiexec/mpiexec.c]: I guess we can use SIGALRM for timeout and for checkpoint together. Just set alarm as minimum of timeout or checkpoint value and set for example some flags.

File: src/pm/hydra2/mpiexec/mpiexec.c (hunk 2 above)
Comment (RozanovAnatoliy) [reply src/pm/hydra2/mpiexec/mpiexec.c]: Done

File: src/pm/hydra2/mpiexec/mpiexec.c (hunk 1 above)
Comment (RozanovAnatoliy) [reply src/pm/hydra2/mpiexec/mpiexec.c]: Fixed

File: src/pm/hydra2/libhydra/timeout/hydra_timeout.c [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    return (timeout->signal > 0) ? timeout->signal : SIGKILL;
+}
+
+int HYD_get_time_left(struct timeout_s *timeout)
+{
+    if (timeout->start_time <= 0) {
+        return -1;
+    }
+
+    int time_diff = (int) (get_current_time() - timeout->start_time);
```
[the remark this answers is not in the document]
Comment (RozanovAnatoliy) [reply src/pm/hydra2/libhydra/timeout/hydra_timeout.c]: Done

**Comment by RozanovAnatoliy** [comment]
@raffenet please review again. I fixed your comments.

**Comment by raffenet** [comment]
> @raffenet please review again. I fixed your comments.

Thanks, I'm going to merge this and work on getting more regular testing on Hydra2.

**Pull request merged**